### DISCUSSION TITLE
Visual cargo management (more reasons for it)

### DISCUSSION THREAD

**Discussion by ChuckStarchaser** [endless-sky/endless-sky#7525] [Ideas]
I began to write this as a bug report post, but then changed my mind, as there is no specific bug to report;  it's just that the whole way cargo is implemented is "hackish code", I suspect.  My present example of it is that I land on a planet with my whole fleet.  I had just been fighting pirates and my fleet was picking up cargo like there's no tomorrow;  so after landing I went to Trading and sold everything there, because I was going to leave with my LEADER SHIP only and leave the rest of the ships parked, and I wanted nothing in my cargo.  But as soon as I tried to depart I get the message that if I leave now, my outfits are going to be sold, as they don't fit in my ship ... ... Uh?  What outfits?  I look in Trading and it says I have 23 tons of outfits in my 22-ton cargo LEADER SHIP, and -1 tons free.  All the other ships are empty, but it forces whatever outfits those are into my poor 22 ton space.  Darn! The worst part is that the planet does not have an Outfitter, so I have no way to know what am I dealing with.  Anything expensive and important?, or just stupid ammo and rockets?  I donno.
  Visual cargo management should actually be available in any planet, regardless of whether it has an Outfitting room, as an outfitting room is basically like an auto-mechanic shop for space-ships.  You should not need a mechanic just to move a bag from the trunk of your MachE to the frunk of your Model 3.  Anywhere you can stop you should be able to do that.
  I think that the problem is that perhaps original Outfits and Cargo were separate domains in the code, and then the code was hacked to make it possible to move outfits to cargo.  Probably needs a fresh re-think, and a code refactoring.
  Here's an idea:  Make marketable categories, like food, metals, etc, be actual containers of marketable items;  so that you can go INTO foods, and see sugar, potatoes, tofu, etc.;  and INTO metals and see iron, copper, aluminum, molybdenum...  Make it so that the category can be traded, so that you don't HAVE TO go into details, but so that you CAN if you want to.  Create a category called Ship Outfits.  Then any outfits bought, or found in space, can go into this marketable category.  Then, you can be in a planet without an Outfitting room, and still see what outfits you have in cargo just by going into Trading, and clicking into the Ship Outfits category.
  The cargo management part that I suggest would be off the main menu when you land at a base, through a Fleet button that's always there in every planet regardless of whether there is a Shipyard or an Outfitters.
  In the Fleet screen you see your fleet as the outlines of your ships, but each ship is as if it were a glass bottle with a level of red liquid indicating the % of fill of its cargo hold.  At the bottom of the red fill, there could be a blue fill level indicating outfits.  Alternatively, there could be one color for each category of cargo, going like a rainbow from red for food to indigo for outfits.
  Here you can click on a category in one ship and drag it to another, or to a trash-can if you want to trash it.
  There would be buttons for Parking and Un-parking.
  Double-clicking a ship would expand it and allow to control weapon placements.
  In other words, it would replace the Ships part of the Info screen, as well as allow cargo reshuffling.
  Question:  The main screen when landing on a planet, is that a planetary screen, rather than a personal device?  Is that why Info is separate from it, by pressing [I] ?  If so, cargo management could be part of Info, instead of from a button in the main screen.  But if so, also, the fact that the main screen is planetary and the Info is personal needs to be made more clear to the player, somehow.  Calling it "Personal", accessible through [P] would be a start.  The old Privateer solution was that by pressing C you called up your (personal, tablet-) Computer, which then filled your screen, and had an LCD screen and buttons you could press.  That made it perfectly clear what was planetary and what was personal.  Planetary info was displayed on green looking CRT screens, as were popular back in those days.  There was no way of not knowing what was public from what was personal;  it was totally obvious.

**Comment by mOctave** [comment]
Selling all just sells commodities. If you click it again, you should be able to sell outfits. You should be able to press [I] and navigate to your flagship view to see what outfits you have in your cargo. I'm not sure if you can selectively sell/store outfits if there isn't an outfitter, but you can still check to see if there's something important that you should keep, or just, as you say, ammunition.

**Reply by ChuckStarchaser** [reply]
> Selling all just sells commodities. If you click it again, you should be able to sell outfits.

Yes, I know this;  and by the way, there's another quasi-bug/annoyance with this Sell button changing to Sell Outfits, in that it seldom happens to me that I try to click the mouse but accidentally double-click.  The sell outfits button should be placed besides the sell-all button;  not "behind" it.

> You should be able to press [I] and navigate to your flagship view to see what outfits you have in your cargo.

Ahh, this I did not know.  Thanks!

**Comment by vitalchip** [comment]
When ships land they're fully unloaded, that's cargo, outfits and passengers. That makes all those things available for trading/outfitting etc. Think of it like putting everything into a holding pen on the dockside.
. 
It would be an easy task to have ships NOT unload when they land but this would need to be set before landing either globally or on a ship by ship basis but then nothing from those ships would be made available for sale/outfitting etc. I suppose there could be a manual/auto option with the secondary option to go ship by ship and unload but it sounds rather tedious to catch edge cases.

If you find yourself in a similar situation again you can use this work around. Unpark enough of your escorts to carry the cargo you have and leave the flagship empty, depart and then press 'H' to hold the fleet then jump to another system and land. Park your escorts.

They will park in the system you just left with cargo intact, you can either go back for them or just unpark them at a later time and they'll come to you with the cargo.

**Reply by ChuckStarchaser** [reply]
> ... Think of it like putting everything into a holding pen on the dockside.

Gottcha.

> Unpark enough of your escorts to carry the cargo you have and leave the flagship empty, depart and then press 'H' to hold the fleet then jump to another system and land. Park your escorts.

I'm ashamed I didn't think of it.  That's brilliant!  Thanks!
But I still think there should be a visual cargo management interface.

Sorry I haven't been around much this past week;  I started working on a planet landed spaceport image to present as example of my other suggestion, but I got stuck at the actual spaceport building.  I don't like the way it looks, but I'm not sure how to improve it or to change it to what.  I'll come out of it ...

**Comment by vitalchip** [comment]
I can certainly imagine a number of cases where some finer control of cargo loading would be an advantage, anything from deliberately putting high value cargo onto the flagship to spreading the load around a fleet to keep ship performance broadly equal.

The problem as I see it is that while Endless Sky holds a lot of appeal for me, it's a small, simple game.

That's not meant as an insult but it's a recognition of how it was designed to be. It's mostly small and 'arcade' and I prefer a bit of nuance and complexity, something more grand and maybe even 'epic'

Take my suggestion above for spreading the load around ships.. why?  Every jump in every ship with every drive type takes a day and your ship arrives within spitting distance of the planet, so there isn't actually any need or use for the nuance of load balancing because it really won't make any substantial difference either way.

Having control over cargo (likewise fighter loading and other similar issues) would be nice but the foundations of the game make it a pointless addition.

I used to play around with some concepts from time to time but I haven't bothered for a while now, the truth is that in order to create the sort of experience I'm looking for some things will need to be made "worse". Ships need to be slower (to create diverse fleets and give players the time to use strategy)  Trading needs to be less predictable (I think it's more immersive to have variable availability, yes it can be frustrating to have to change plans when somethingi sn't available or they don't have as many as you might want but it's more like trading than just bouncing back and forth between planets buying ever larger quantities of the same items without end)

However, this isn't the direction the game is going, as was pointed out to me on Discord "this is an arcade game". Far too many people are content to shout that they "want better Ai" or "xyz needs improved" but the same people refuse to let go of the arcade underpinnings and can't understand the two are in conflict.

On the cargo front, I changed the shipinfo screen a while ago, it's not a way to load anything but it is perhaps a clearer way of seeing what is in the ships. When 'dockside' it lists everything as "pooled cargo", when in flight as "ship cargo" and when parked as (surprise) "parked cargo".  So you get lists like this, I thought it was better.


<img width="274" alt="Screenshot 2022-11-20 at 15 06 10" src="https://user-images.githubusercontent.com/65342577/203015565-f976d1d8-08fd-4dfd-8db7-557e6507b27a.png">

<img width="273" alt="Screenshot 2022-11-20 at 15 15 33" src="https://user-images.githubusercontent.com/65342577/203015653-e28c9359-82a5-434a-9c25-15b5c6981c86.png">
<img width="270" alt="Screenshot 2022-11-20 at 15 15 50" src="https://user-images.githubusercontent.com/65342577/203015667-00266f3b-7bba-4a86-b003-f462b860657f.png">

<img width="280" alt="Screenshot 2022-11-20 at 19 14 50" src="https://user-images.githubusercontent.com/65342577/203015694-fd135dbf-17af-469e-9bfd-85552fdea8e6.png">

**Comment by ChuckStarchaser** [comment]
@vitalchip You're very right;  and you took words out of my mouth that I didn't even know were there.  Indeed, I frankly hate arcades.  I continue to play this game because I want to get through the rest of the story-line.  If it was up to me, I'd slow down and complexify everything, then throw in features like being able to go many more places when landed than just the same 5 or 6 at every planet.  I'd have trains, and trips to other cities, and stories that take you on such planet-side journeys, as was implemented in Privateer 2.  I'd make space-fighting at least as slow and complex as it was in Homeworld, and would make ship design and specialization the inescapable solution by multiplying the number of available outfits without increasing outfit space, which would result in the types of space battle complexities of Masters of Orion 2.
  Zitchas was telling me of an experimental version of the game that was once put together to test many ideas, and he described space-battles as slow and "cinematic";  and I fell in love with that version without even having experienced it myself.
  But if the powers that be want this to remain an arcade, an arcade it will remain, I guess.  Sad, but might as well get over it.
  Back to the cargo question, in Privateer 2 when you hired a cargo ship (there was no way to "own" one, but you could hire one), you'd see where on the ship everything went;  but for the life of me I cannot remember why it was important to decide where everything went.  I only remember that it took some time, sometimes, to get it all optimized.
  Nah, I went looking for a video to remind myself, and all it was is that it made it easier to trade if one compartmentalized the cargo by categories, as the way to maximize profit, in that game, was to keep some of the cargo across several trade route hops.  Which is another annoyingly simple characteristic of trade in this game that the optimal way to trade is to aways load up on one thing, and dump everything at the next stop.
  Anyways, for the curious, here's a video of Privateer 2:
https://youtu.be/O6aGapGXuag
  And here's a video of Masters of Orion II:
https://youtu.be/Ys6ir_IDR1c
  And here's a video of Homeworld:
https://youtu.be/4l0SiXyNd6Q

**Comment by SeanScherer** [comment]
Interesting vitalchip and ChuckStarchaser, cause this is pretty much the same feeling I have (I like the game, but its the story, trading and missions that I find attractive. The in-flight setup is more something I "tolerate" to have the rest :P ).

Been thinking about which directions it might be interesting to spin things. I'll let you know if I have any more concrete ideas!
